api/v1alpha1/executionhook_types.go

```diff
@@ -0,0 +1,186 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// ExecutionHookSpec defines the desired state of ExecutionHook
+	Spec ExecutionHookSpec `json:"spec,omitempty"`
+	// ExecutionHookStatus defines the observed state of ExecutionHook
+	Status *ExecutionHookStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ExecutionHookList contains a list of ExecutionHook
+type ExecutionHookList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ExecutionHook `json:"items"`
+}
+
+// ExecutionHookSpec defines the desired state of ExecutionHook
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Application Snapshot Controller.
+type ExecutionHookSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// PodSelection defines how to select pods and containers to run
+	// the executionhook. If multiple pod/containers are selected, the action will executed on them
+	// asynchronously. If execution ordering is required, caller has to implement the logic and create
+	// different hooks in order.
+	// This field is required.
+	PodSelection PodSelection `json:"podSelection"`
+
+	// Name of the HookAction. This is required.
+	ActionName string `json:"actionName"`
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where exactly one of them must be defined so that the hook controller knows where to
+// run the hook.
+// TODO: Add validation to validation ensure exactly one of them are defined.
+type PodSelection struct {
+	// PodContainerNamesList lists the pods/containers on which the ExecutionHook
+	// should be executed. If not specified, the ExecutionHook controller will find
+	// all pods and containers based on PodContainerSelector.
+	// If both PodContainerNamesList and PodContainerSelector are not
+	// specified, the ExecutionHook cannot be executed and it will fail.
+	// +optional
+	PodContainerNamesList []PodContainerNames `json:"podContainerNamesList,omitempty"`
+
+	// PodContainerSelector is for hook controller to find pods and containers
+	// based on the pod label selector and container names
+	// If PodContainerNamesList is specified, this field will not be used.
+	// +optional
+	PodContainerSelector *PodContainerSelector `json:"podContainerSelector,omitempty"`
+}
+
+// PodContainerNames lists the containers the ExecutionHook should be executed on in a Pod.
+type PodContainerNames struct {
+	// This field is required
+	PodName string `json:"podName"`
+
+	// +optional
+	// If empty, hook action will be executed on all containers in the pod.
+	ContainerNames []string `json:"containerNames,omitempty"`
+}
+
+// PodContainerSelector defines the selector and containers the ExecutionHook
+// should be executed on.
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
```

Comment:
Add "If not specified, all pods will be selected".

Reply:
From the KEP, looks like this has to be specified?
```
        // PodSelection defines how to select pods and containers to run
	// the executionhook. If multiple pod/containers are selected, the action will executed on them
	// asynchronously. If execution ordering is required, caller has to implement the logic and create
	// different hooks in order.
	// This field is required.
        PodSelection PodSelection
```

Reply:
Sorry, my bad. Mistook it for a different field. Ignore what I said.

Reply:
@xing-yang  Should we add this to the KEP if we want to select all pods.

Reply:
I've added a comment. I don't know if this was discussed before. But, I am not sure if we want to select all pods.

Reply:
Correct.  PodSelection must be specified so one of the two choices (PodContainerNamesList and PodContainerSelector) must be selected.  Let's assume that PodContainerSelector is specified. That just means it is not nil so it can still be an empty struct. Inside PodContainSelector, both PodSelector and ContainerList are optional.  If PodSelector is not specified but ContainerList is specified (not specified means all containers), how do we know which pod the container is in?  We still need to go through the pods to find out which pod has this container.  Maybe instead of saying "all pods will be selected", you can explain briefly how would a pod be chosen if PodSelector is not specified?

Reply:
So that means, if an empty `PodContainerSelector`  is used, then we will be running the hook on all containers in all pods.

Reply:
Pushed changes to resolve this.